feature-terminate-insurance/src/main/kotlin/com/hedvig/android/feature/terminateinsurance/data/TerminateInsuranceRepository.kt

```diff
@@ -0,0 +1,56 @@
+package com.hedvig.android.feature.terminateinsurance.data
+
+import com.apollographql.apollo3.ApolloClient
+import com.hedvig.android.apollo.OperationResult
+import com.hedvig.android.apollo.safeExecute
+import com.hedvig.android.feature.terminateinsurance.InsuranceId
+import kotlinx.datetime.toKotlinLocalDate
+import octopus.FlowTerminationDateNextMutation
+import octopus.FlowTerminationStartMutation
+import octopus.type.FlowTerminationDateInput
+import java.time.LocalDate
+
+internal class TerminateInsuranceRepository(
+  private val apolloClient: ApolloClient,
+) {
+  private var terminationContext: Any? = null
+
+  suspend fun startTerminationFlow(insuranceId: InsuranceId): TerminationStep {
+    return when (val res = apolloClient.mutation(FlowTerminationStartMutation(insuranceId.id)).safeExecute()) {
+      is OperationResult.Error -> TerminationStep.Failed(res.message)
+      is OperationResult.Success -> {
+        terminationContext = res.data.flowTerminationStart.context
+        res.data.flowTerminationStart.currentStep.toTerminationStep()
+      }
+    }
+  }
+
+  suspend fun setTerminationDate(terminationDate: LocalDate): TerminationStep {
+    val nextMutation = FlowTerminationDateNextMutation(
+      context = terminationContext!!,
+      input = FlowTerminationDateInput(terminationDate.toKotlinLocalDate()),
+    )
+    return when (val res = apolloClient.mutation(nextMutation).safeExecute()) {
+      is OperationResult.Error -> TerminationStep.Failed(res.message)
+      is OperationResult.Success -> {
+        terminationContext = res.data.flowTerminationDateNext.context
+        res.data.flowTerminationDateNext.currentStep.toTerminationStep()
+      }
+    }
+  }
+}
+
+private fun FlowTerminationStartMutation.Data.FlowTerminationStart.CurrentStep.toTerminationStep(): TerminationStep = when (this) {
+  is FlowTerminationStartMutation.Data.FlowTerminationStart.FlowTerminationDateStepCurrentStep -> TerminationStep.Date(minDate, maxDate)
```

Comment:
Would suggest importing the entire thing here so we only see the last type, not everything that comes before it. This might also be easier after we introduce the fragments

Reply:
Fragments does not help here unfortunately... apollo still generates two different implementations, one in `FlowTerminationStartMutation` and one in `FlowTerminationDateNextMutation`:
```
public sealed interface CurrentStep : StepFragment {
        public override val __typename: String

        public companion object {
          public fun CurrentStep.stepFragment() = this as? StepFragment

          public fun CurrentStep.asFlowTerminationDateStep() = this as?
              FlowTerminationDateStepCurrentStep

          public fun CurrentStep.asFlowTerminationFailedStep() = this as?
              FlowTerminationFailedStepCurrentStep

          public fun CurrentStep.asFlowTerminationSuccessStep() = this as?
              FlowTerminationSuccessStepCurrentStep
        }
      }
```

I fixed the imports though, so its a bit easier on the eyes.